src/twisted/conch/ssh/transport.py

```diff
@@ -1634,31 +1660,46 @@ def _ssh_KEX_ECDH_REPLY(self, packet):
 
         @return: A deferred firing when key exchange is complete.
         """
+
         def _continue_KEX_ECDH_REPLY(ignored, hostKey, pubKey, signature):
             # Save off the host public key.
             theirECHost = hostKey
 
-            # Take the provided public key and transform it into a format
-            # for the cryptography module
-            theirECPub = ec.EllipticCurvePublicNumbers.from_encoded_point(
-                            self.curve, pubKey).public_key(
-                            default_backend())
+            if self.kexAlg.find(b'curve25519') >= 0:
+                theirECPub = nacl.public.PublicKey(pubKey)
 
-            # We need to convert to hex,
-            # so we can convert to an int
-            # so we can make a multiple precision int.
-            sharedSecret = MP(
-                           int(
-                           binascii.hexlify(
-                             self.ecPriv.exchange(ec.ECDH(), theirECPub)), 16))
+                # Compute the shared secret
+                sharedSecret = MP(
+                    int(binascii.hexlify(
+                        crypto_scalarmult(self.ecPriv._private_key,
+                                          theirECPub._public_key)),
+                        16))
+
+                encPub = self.ecPub._public_key
+            else:
+                # Take the provided public key and transform it into a format
+                # for the cryptography module
+                theirECPub = ec.EllipticCurvePublicNumbers.from_encoded_point(
+                                self.curve, pubKey).public_key(
+                                default_backend())
+
+                # We need to convert to hex,
```

Comment:
Given that this comment is repeated twice, and now the `MP(...` expression is repeated 4 times, some refactoring seems to be in order.